specs/directory.md

```diff
@@ -0,0 +1,387 @@
+| `CACHE`   | Directory for user-specific cache                                                   |
+
+Although it is recommended to install `notation` with its plugins and default configurations at the system level, it is possible to install at the user level.
+
+On Unix systems, `notation` follows [Filesystem Hierarchy Standard][FHS] for system level directories and [XDG Base Directory Specification][XDG] for user level directories. On Windows, [Known Folders][KF] and [App Settings][AS] are followed equivalently. On Darwin, [macOS File System][macOS_FS] with [System Integrity Protection][SIP] is followed equivalently. If a file with the same name exists at the system level and the user level, the file at the user level takes over the priority.
+
+### System Level
+
+Default directory paths for various operating systems at system level are specified as below.
+
+| Directory | Unix           | Windows                       | Darwin                         |
+| --------- | -------------- | ----------------------------- | ------------------------------ |
+| `BIN`     | `/usr/bin`     | `%ProgramFiles%/notation/bin` | `/usr/local/bin`               |
+| `LIBEXEC` | `/usr/libexec` | `%ProgramFiles%`              | `/usr/local/lib`               |
+| `CONFIG`  | `/etc`         | `%ProgramData%`               | `/Library/Application Support` |
+
+`CACHE` is omitted since it is user specific.
+
+### User Level
+
+Default directory paths for various operating systems at user level are specified as below.
+
+| Directory | Unix               | Windows          | Darwin                          |
+| --------- | ------------------ | ---------------- | ------------------------------- |
+| `LIBEXEC` | `$XDG_CONFIG_HOME` | `%AppData%`      | `~/Library/Application Support` |
+| `CONFIG`  | `$XDG_CONFIG_HOME` | `%AppData%`      | `~/Library/Application Support` |
+| `CACHE`   | `$XDG_CACHE_HOME`  | `%LocalAppData%` | `~/Library/Caches`              |
+
+There is no default `BIN` path at user level since the `notation` binary can be put anywhere as long as it in the `PATH` environment variable. Common directories on Unix/Darwin are `~/bin` and `~/.local/bin` where manual `PATH` update by users may be required.
+
+## Structure
+
+The overall directory structure for `notation` is summarized as follows.
+
+```
+{BIN}
+└── notation
+{CACHE}
+└── notation
+    └── signature
+        └── {manifest-digest-algorithm}
+            └── {manifest-digest}
+                └── {signature-digest-algorithm}
+                    └── {signature-digest}.sig
+{CONFIG}
+└── notation
+    ├── config.json
+    ├── signing
+    │   ├── certificates
+    │   │   └── {key-name}.crt
+    │   ├── keys
+    │   │   └── {key-name}.pem
+    │   └── keys.json
+    └── trust
+        ├── policy.json
+        └── store
+            └── {trust-store-type}
+                └── {named-store}
+                    └── {cert-file}

```

Comment:
I feel these paths are bit lengthy to traverse manually. Can we use shorter ones like below?
```
{CONFIG}/notation/signing-keys/{key-name}/cert.pem
{CONFIG}/notation/signing-keys/{key-name}/key.pem
{CONFIG}/notation/trust-store/{trust-store-type}/*
{CONFIG}/notation/trust-policy.json
```

I personally feel shorter paths are easier to use. Would like to hear others opinions. @gokarnm what do you think?

Reply:
I don't have strong opinions, but current dir structure has some additional levels due to `notation\signing` and `notation\trust`. @shizhMSFT is there a benefit to having these sub directories, anything we loose by flattening the structure?

 Things to consider

1. Users will initially have to be aware of some paths the dir structure when we don't provide them CLI command (mainly trust policy and trust store), so we can keep paths short as long as they are clear and unambiguous.
2. Administrators who will deploy trust stores and policies will not use CLI commands, they will copy files in specific locations. 

I like the suggestion of combining the `signing/certificates` and `signing/keys` to `signing-keys/{key-name}`. My suggestions.

```
{CONFIG}/notation/signingKeys.json
{CONFIG}/notation/trustPolicy.json
{CONFIG}/notation/config.json
{CONFIG}/notation/signing-keys/{key-name}/cert.crt 
{CONFIG}/notation/signing-keys/{key-name}/key.pem 
{CONFIG}/notation/trust-store/{trust-store-type}/* 
```

Reply:
`trust/store` and `trust/policy` are better than `trust-store` and `trust-policy` as they are all under `trust` and are well organized. It also avoids naming convention. For example, `signing-keys` and `signingKeys` are not consistency in terms of naming.

Reply:
> Administrators who will deploy trust stores and policies will not use CLI commands, they will copy files in specific locations.

They will deploy those using scripts, not drag-and-drop GUIs.

Reply:
> trust/store and trust/policy are better than trust-store and trust-policy as they are all under trust and are well organized. 

The path `{CONFIG}/notation/trust` is ambiguous. It exists on the file system but does not provide any real value. The flattened structure is better as each and every possible path is self-explanatory.

Now that I understand better about these files, I propose below structure
```
{CONFIG}/notation/signingkeys.json
{CONFIG}/notation/trustpolicy.json
{CONFIG}/notation/config.json
{CONFIG}/notation/local-keys/{key-name}/cert.pem
{CONFIG}/notation/local-keys/{key-name}/key.pem
{CONFIG}/notation/local-keys/{key-name}/chain.pem 
{CONFIG}/notation/trust-store/{trust-store-type}/* 
```

AFAIK camel case for multi-word config file names (trustPolicy.json, signingKeys.json) is not common (for example `.gitconfig` or `.gitignore`). I tried to look for any examples in docker but couldn't find any.

Reply:
The folder name `local-keys` seems making more sense than `signing-keys`.

BTW, should we remove all hyphens? so we have `truststore` and `localkeys`?